machine_learning/loss_functions/categorical_cross_entropy.py

```diff
@@ -0,0 +1,58 @@
+"""
+Categorical Cross-Entropy Loss
+
+This function calculates the Categorical Cross-Entropy Loss between true class
+labels and predicted class probabilities.
+
+Formula:
+Categorical Cross-Entropy Loss = -Σ(y_true * log(y_pred))
+
+Resources:
+- [Wikipedia - Cross entropy](https://en.wikipedia.org/wiki/Cross_entropy)
+"""
+
+import numpy as np
+
+
+def categorical_crossentropy(
+    y_true: np.ndarray, y_pred: np.ndarray, epsilon: float = 1e-15
+) -> float:
+    """
+    Calculate Categorical Cross-Entropy Loss between true class labels and
+    predicted class probabilities.
+
+    Parameters:
+    - y_true: True class labels (one-hot encoded) as a NumPy array.
+    - y_pred: Predicted class probabilities as a NumPy array.
+    - epsilon: Small constant to avoid numerical instability.
+
+    Returns:
+    - ce_loss: Categorical Cross-Entropy Loss as a floating-point number.
+
+    Example:
+    >>> true_labels = np.array([[1, 0, 0], [0, 1, 0], [0, 0, 1]])
+    >>> pred_probs = np.array([[0.9, 0.1, 0.0], [0.2, 0.7, 0.1], [0.0, 0.1, 0.9]])
+    >>> categorical_crossentropy(true_labels, pred_probs)
+    0.18913199175146167
+
+    >>> y_true = np.array([[1, 0], [0, 1]])
+    >>> y_pred = np.array([[0.9, 0.1, 0.0], [0.2, 0.7, 0.1]])
+    >>> categorical_crossentropy(y_true, y_pred)
+    Traceback (most recent call last):
+        ...
+    ValueError: Input arrays must have the same length.
+    """
+    if y_true.shape != y_pred.shape:
+        raise ValueError("Input arrays must have the same length.")
+
+    # Clip predicted probabilities to avoid log(0)
+    y_pred = np.clip(y_pred, epsilon, 1 - epsilon)
```

Comment:
Also, since `y_pred` represents class probabilities, shouldn't it be normalized so that it sums to 1? We should either enforce this with a ValueError or automatically normalize input vectors ourselves.

Reply:
I clipped the upper bound to 1 - epsilon to completely avoid the case of log(0), as a probability of 1 may lead to log(0)
Also added a check for y_pred so that it sums to 1

Reply:
After thinking about it for a bit, I think it should be fine to clip the upper bound at just 1. Given the formula `-Σ(y_true * log(y_pred))`, the only time when we'd get a $\log(0)$ in the formula is if `y_pred` is 0. If `y_pred` is 1, then we get $\log(1) = 0$ in the term, which isn't a problem.

Reply:
Oh Yes, you are correct. I don't know how I missed that. 
Thanks